src/identity_tree.rs

```diff
@@ -1,26 +1,247 @@
+
+impl TreeVersion {
+    pub async fn peek_next_update(&self) -> Option<TreeUpdate> {
+        let data = self.0.read().await;
+        data.peek_next_update().await
+    }
+
+    pub async fn apply_next_update(&self) {
+        let mut data = self.0.write().await;
+        data.apply_next_update().await;
+    }
+
+    pub async fn update(&self, leaf_index: usize, element: Hash) {
+        let mut data = self.0.write().await;
+        data.update(leaf_index, element);
+    }
+
+    pub async fn next_version(&self) -> Self {
+        let mut data = self.0.write().await;
+        data.next_version()
+    }
+
+    pub async fn append_many_fresh(&self, updates: &[TreeUpdate]) {
+        let mut data = self.0.write().await;
+        let next_leaf = data.next_leaf;
+        updates
+            .iter()
+            .filter(|update| update.leaf_index >= next_leaf)
+            .for_each(|update| {
+                data.update(update.leaf_index, update.element);
+            });
+    }
+
+    pub async fn next_leaf(&self) -> usize {
+        let data = self.0.read().await;
+        data.next_leaf
+    }
+
+    async fn get_proof(&self, leaf: usize) -> (Hash, Proof) {
+        let tree = self.0.read().await;
+        (
+            tree.tree.root(),
+            tree.tree
+                .proof(leaf)
+                .expect("impossible, tree depth mismatch between database and runtime"),
+        )
+    }
+}
+
+pub struct TreeItem {
+    pub status:     Status,
+    pub leaf_index: usize,
+}
+
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Status {
+    Pending,
+    Mined,
+}
+
+#[derive(Debug, Error)]
+#[error("unknown status")]
+pub struct UnknownStatus;
+
+impl FromStr for Status {
+    type Err = UnknownStatus;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "pending" => Ok(Self::Pending),
+            "mined" => Ok(Self::Mined),
+            _ => Err(UnknownStatus),
+        }
+    }
+}
+
+impl From<Status> for &str {
+    fn from(scope: Status) -> Self {
+        match scope {
+            Status::Pending => "pending",
+            Status::Mined => "mined",
+        }
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct InclusionProof {
+    pub status: Status,
+    pub root:   Field,
+    pub proof:  Proof,
+}
+
+#[derive(Clone)]
+pub struct TreeState {
+    mined:  TreeVersion,
+    latest: TreeVersion,
+}
 
 impl TreeState {
     #[must_use]
-    pub fn new(tree_depth: usize, initial_leaf: Field) -> Self {
-        Self {
-            next_leaf:   0,
-            merkle_tree: PoseidonTree::new(tree_depth, initial_leaf),
+    pub const fn new(mined: TreeVersion, latest: TreeVersion) -> Self {
+        Self { mined, latest }
+    }
+
+    #[must_use]
+    pub fn get_latest_tree(&self) -> TreeVersion {
+        self.latest.clone()
+    }
+
+    #[must_use]
+    pub fn get_mined_tree(&self) -> TreeVersion {
+        self.mined.clone()
+    }
```

Comment:
I'd rename these to something that indicates that it's a clone under the hood. It confused me at the start until I looked at the impl.

Reply:
It's a clone of an `Arc`, why would you care? I'm assuming if someone is reading these implementations, they need to understand the internal types. From the outside, you just get a reference to the tree.

Reply:
Ah, I missed that it was an `Arc`! All good.